path/path.go

```diff
@@ -3,187 +3,237 @@ package path
 )
 
-// A Path represents an ipfs content path:
-//   - <cid>/path/to/file
-//   - /ipfs/<cid>
-//   - /ipns/<cid>/path/to/folder
-//   - etc
-type Path string
+type Namespace string
 
-// ^^^
-// TODO: debate making this a private struct wrapped in a public interface
-// would allow us to control creation, and cache segments.
+func (namespace Namespace) String() string {
+	return string(namespace)
+}
 
-// FromString safely converts a string type to a Path type.
-func FromString(s string) Path {
-	return Path(s)
+// Mutable returns false if the data under this namespace is guaranteed to not change.
+func (namespace Namespace) Mutable() bool {
+	return namespace != IPFSNamespace && namespace != IPLDNamespace
 }
 
-// FromCid safely converts a cid.Cid type to a Path type.
-func FromCid(c cid.Cid) Path {
-	return Path("/ipfs/" + c.String())
+const (
+	IPFSNamespace Namespace = "ipfs"
+	IPNSNamespace Namespace = "ipns"
+	IPLDNamespace Namespace = "ipld"
+)
+
+// Path is a generic, valid, and well-formed path. A valid path is shaped as follows:
+//
+//	/{namespace}/{root}[/remaining/path]
+//
+// Where:
+//
+//  1. Namespace is "ipfs", "ipld", or "ipns".
+//  2. If namespace is "ipfs" or "ipld", "root" must be a valid [cid.Cid].
+//  3. If namespace is "ipns", "root" may be a [ipns.Name] or a [DNSLink] FQDN.
+//
+// [DNSLink]: https://dnslink.dev/
+type Path interface {
+	// String returns the path as a string.
+	String() string
+
+	// Namespace returns the first component of the path. For example, the namespace
+	// of "/ipfs/bafy" is "ipfs".
+	Namespace() Namespace
+
+	// Segments returns the different elements of a path delimited by a forward
+	// slash ("/"). The returned array must not contain any empty segments, and
+	// must have a length of at least two: the first element must be the namespace,
+	// and the second must be root.
+	//
+	// Examples:
+	// 		- "/ipld/bafkqaaa" returns ["ipld", "bafkqaaa"]
+	// 		- "/ipfs/bafkqaaa/a/b/" returns ["ipfs", "bafkqaaa", "a", "b"]
+	// 		- "/ipns/dnslink.net" returns ["ipns", "dnslink.net"]
+	Segments() []string
 }
 
-// Segments returns the different elements of a path
-// (elements are delimited by a /).
-func (p Path) Segments() []string {
-	cleaned := path.Clean(string(p))
-	segments := strings.Split(cleaned, "/")
+var _ Path = path{}
 
-	// Ignore leading slash
-	if len(segments[0]) == 0 {
-		segments = segments[1:]
-	}
+type path struct {
+	str       string
+	namespace Namespace
+}
+
+func (p path) String() string {
+	return p.str
+}
+
+func (p path) Namespace() Namespace {
+	return p.namespace
+}
 
-	return segments
+func (p path) Segments() []string {
+	return StringToSegments(p.str)
 }
 
-// String converts a path to string.
-func (p Path) String() string {
-	return string(p)
+// ImmutablePath is a [Path] which is guaranteed to have an immutable [Namespace].
+type ImmutablePath interface {
+	Path
+
+	// Cid returns the [cid.Cid] of the root object of the path.
+	Cid() cid.Cid
```

Comment:
:warning:  The name makes is not obvious what this is.
For `/ipfs/cid/dir/terminus` Could be  CID of the terminating element.

Let's make it clear this is the root one:

```suggestion
	// RootCid returns the [cid.Cid] of the root identifier of the path.
	RootCid() cid.Cid
```